scripts/mirror-checker.py

```diff
@@ -1,77 +1,500 @@
 
 import argparse
+import concurrent.futures
 import os
 import shlex
 import subprocess
 import sys
 import tempfile
+from datetime import datetime, timezone
+from typing import TextIO
+from pathlib import Path
+from typing import Any
 
+import dns.resolver
+from prettytable import PrettyTable
+from pycountry import countries
 
-def run_cmd(cmd: list[str]) -> int:
+
+def run_cmd(cmd: list[str], output: TextIO = sys.stdout) -> int:
     """Run a cmd list. Return output."""
     # Print out command bash -x style
-    print(f"+ {shlex.join(cmd)}")
+    print(f"+ {shlex.join(cmd)}", file=output)
     with subprocess.Popen(
         cmd,
         stdout=subprocess.PIPE,
         stderr=subprocess.PIPE,
         encoding="utf-8",
     ) as proc:
         try:
-            for line in iter(proc.stdout.readline, ""):
-                print(line, end="")
+            for line in iter(proc.stdout.readline, ""):  # type: ignore
+                print(line, end="", file=output, flush=True)
         except KeyboardInterrupt:
-            print("Killed by user.")
+            print("Killed by user.", file=output, flush=True)
         finally:
-            proc.stdout.close()
+            proc.stdout.close()  # type: ignore
             proc.wait()
 
         return proc.returncode
 
 
-def parse_args() -> argparse.Namespace:
+def check_mirror(
+    mirror_url: str,
+    release: str,
+    output: TextIO = sys.stdout,
+) -> int:
+    """Check the health of a single mirror for a particular release.
+
+    Uses debootstrap to create a schroot using the specified mirror_url.
+    Returns a 0 on success and a 1 on failure.
+    """
+    # Default to 1 so if run_cmd fails then we still get a failure.
+    ret = 1
+    with tempfile.TemporaryDirectory() as tempdir:
+
+        print(datetime.now(timezone.utc), file=output)
+        print(f"Creating schroot in: {tempdir}", file=output)
+        print(f"Targeting release: {release}", file=output)
+        print(f"Using mirror: {mirror_url}", file=output)
+
+        debootstrap_cmd = ["debootstrap", release, tempdir, mirror_url]
+        ret = run_cmd(debootstrap_cmd, output=output)
+
+    return ret
+
+
+def check_country_mirror(
+    country_code: str,
+    release: str,
+    output: TextIO = sys.stdout,
+) -> int:
+    """Check the health of a single country mirror for a particular release.
+
+    Uses debootstrap to create a schroot using the specified mirror. Returns
+    a 0 on success and a 1 on failure.
+    """
+    mirror: str = f"http://{country_code}.archive.ubuntu.com/ubuntu/"
+    return check_mirror(mirror, release, output=output)
+
+
+def check_primary_archive(
+    release: str,
+    output: TextIO = sys.stdout,
+) -> int:
+    mirror: str = "http://archive.ubuntu.com/ubuntu/"
+    return check_mirror(mirror, release, output=output)
+
+
+def get_cname(domain: str) -> str:
+    """Get CNAME records for domain."""
+    try:
+        answers = dns.resolver.Resolver().resolve(domain, "CNAME")
```

Comment:
suggestion: https://serverfault.com/a/574077 is claiming that it's technically possible but out of spec.  what about:
```suggestion
        [answers] = dns.resolver.Resolver().resolve(domain, "CNAME")
```
which means a return from `resolve()` with more than 1 element will `raise`.

Reply:
Yeah I like this better. Then we could also report information about poorly configured mirrors too.